engine/src/main/java/org/terasology/physics/bullet/shapes/BulletCollisionShapeFactory.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2017 MovingBlocks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.physics.bullet.shapes;
+
+import org.terasology.math.geom.Vector3f;
+import org.terasology.physics.shapes.BoxShape;
+import org.terasology.physics.shapes.CollisionShapeFactory;
+import org.terasology.physics.shapes.CompoundShape;
+import org.terasology.physics.shapes.ConvexHullShape;
+import org.terasology.physics.shapes.SphereShape;
+
+import java.util.List;
+
+public class BulletCollisionShapeFactory implements CollisionShapeFactory {
```

Comment:
What's the value of using this type of implementation? This just adds an additional layer of complexity to the implementation. What is the value of this if you are basically going to use this to just declare the object. This implementation just introduces complexity.

Reply:
This applies dependency inversion, the factory hides the implementation details from the used object/interface. You could switch this to any other instance of a CollisionShapeFactory and migrate the entire codebase to a different framework.

Reply:
I understand a factory is used to generalize the implementation so it can be exchanged for something else. The entire model is already tied to how jbullet works. You can interface out the rest of the shapes and generalize it in the fashion or even wrap the objects themselves.